adam_core/utils/mpc.py

```diff
@@ -0,0 +1,488 @@
+    isot_string = "{:d}-{:02d}-{:02d}".format(year, month, day)
+
+    if len(epoch_pf) > 5:
+        fractional_day = float("." + epoch_pf[5:])
+        hours = int((24 * fractional_day))
+        minutes = int(60 * ((24 * fractional_day) - hours))
+        seconds = 3600 * (24 * fractional_day - hours - minutes / 60)
+        isot_string += "T{:02d}:{:02d}:{:09.6f}".format(hours, minutes, seconds)
+
+    return isot_string
+
+
+def _lookup_mpc(x: str) -> int:
+    # Convert the single character dates into integers.
+    try:
+        x = int(x)
+    except ValueError:
+        x = ord(x) - 55
+    if x < 0 or x > 31:
+        raise ValueError
+    return x
+
+
+def convert_mpc_packed_dates(pf_tt: npt.ArrayLike) -> Time:
+    """
+    Convert MPC packed form dates (in the TT time scale) to
+    MJDs in TT. See: https://minorplanetcenter.net/iau/info/PackedDates.html
+    for details on the packed date format.
+
+    Parameters
+    ----------
+    pf_tt : `~numpy.ndarray` (N)
+        MPC-style packed form epochs in the TT time scale.
+
+    Returns
+    -------
+    mjd_tt : `~astropy.time.core.Time` (N)
+        Epochs in TT MJDs.
+    """
+    pf_tt = np.asarray(pf_tt)
+    isot_tt = np.empty(len(pf_tt), dtype="<U32")
+
+    for i, epoch in enumerate(pf_tt):
+        isot_tt[i] = _unpack_mpc_date(epoch)
+
+    epoch = Time(isot_tt, format="isot", scale="tt")
+    return epoch
+
+
+def _pack_numbered_designation(designation: str) -> str:
+    """
+    Pack a numbered MPC designation.
+
+    Examples of numbered designations:
+        Numbered      Packed
+        3202          03202
+        50000         50000
+        100345        A0345
+        360017        a0017
+        203289        K3289
+        620000        ~0000
+        620061        ~000z
+        3140113       ~AZaz
+        15396335      ~zzzz
+
+    Parameters
+    ----------
+    designation : str
+        MPC numbered designation.
+
+    Returns
+    -------
+    designation_pf : str
+        MPC packed numbered designation.
+
+    Raises
+    ------
+    ValueError : If the numbered designation cannot be packed.
+        If the numbered designation is larger than 15396335.
+    """
+    number = int(designation)
+    if number > 15396335:
+        raise ValueError(
+            "Numbered designation is too large. Maximum supported is 15396335."
+        )
+
+    if number <= 99999:
+        designation_pf = "{:05}".format(number)
+    elif (number >= 100000) & (number <= 619999):
+        ind = int(np.floor(number / 10000))
+        designation_pf = "{}{:04}".format(BASE62[ind], number % 10000)
+    else:
+        x = number - 620000
+        number_pf = []
+        while x:
+            number_pf.append(BASE62[int(x % 62)])
+            x = int(x / 62)
+
+        number_pf.reverse()
+        designation_pf = "~{}".format("".join(number_pf).zfill(4))
+
+    return designation_pf
+
+
+def _pack_provisional_designation(designation: str) -> str:
```

Comment:
Why is this private behind a `_`? I'd actually like to use it more than `pack_mpc_designation`. I don't want to accidentally pack a number.

Reply:
I wanted to keep it simple and let `pack_mpc_designation` take care of trying numbered, then provisional, then survey packing. Happy to make them public if it seems like they are going to be used directly.